### ISSUE TITLE
Dark Mode

### ISSUE THREAD

**Issue by bluespidernav** [qtwebkit/qtwebkit#938]
@media (prefers-color-scheme: dark) {
    ...
}

In an embedded QWebPage I'd like CSS selectors such as the above to detect "dark" mode and display differently.

How do I tell the QWebView/QWebPage or its settings that.
I'd first detect whether my app is in dark mode and then when I create the page I need to tell qt webkit. But how?

**Comment by annulen** [comment]
It's not possible currently. Proper Qt implementation of dark mode in qtwebkit-dev is still subject for a discussion, however I'm not sure if backporting this feature to qtwebkit-5.212 makes sense.

**Comment by annulen** [comment]
cc @Emdek

**Comment by bluespidernav** [comment]
Ok understood, my workaround will be to use a data-theme and apply using JS code on the page as in this particular case I have the luxury of controlling the page source! It's perfectly fair that proper support say via QWebSettings would be a low priority feature albeit nice to have.

**Comment by annulen** [comment]
It's somewhat more complicated in practice. Default behavior in 5.212.x and older versions is using QStyle of QWebView widget (or style of QGraphicsView, if you are using QGraphicsWebView), or falls back to style of QApplication. In most real-world applications which don't do custom styling it means that system-wide Qt theme is used. However, there is no easy way to determine if some particular style is "light" or "dark"

One possible approach to this issue is to stop using system Qt styles completely and have two built-in styles for light and dark.

**Comment by Emdek** [comment]
@annulen, we discussed that Qt style part in the past, AFAIR the conclusion was that it would b best to have dedicated style for QtWebKit but allow to replace it.
Light vs dark mode is about two color palettes for single Qt widget style anyway.

**Comment by annulen** [comment]
But we cannot apply arbitrary palettes to arbitrary styles, that may look ugly.

**Comment by Emdek** [comment]
@annulen, tell that to maintainers of color scheme in KDE. ;-)
It's true that some styles might have issues but these are style bugs (but it's obvious who would be blamed for them).
Also, we should default to own widget style **anyway**.

**Comment by Emdek** [comment]
For now could simply default to Fusion.

**Comment by annulen** [comment]
From WebKit side, current situation is following:
* RenderTheme was changed from being per-page object to singleton, which means there cannot be per-page style anymore, only one chosen style with possible "light" and "dark" variation. Previously, we used style of respective QWebView/QGraphicsView. It should be possible to restore old way of things, but I'm not sure if it's really needed.
* In WK2 it's only possible to have one global QStyle, and default behavior is to avoid `WebKit::initializeWebKitWidgets()` call and use "mobile" theme instead (see below)
* Mobile theme has advantage over QStyle-base themes in that all controls like checkboxes or radio buttons are scalable, while QStyle ones have fixed sizes (or may have choice from few size variants).

**Comment by Emdek** [comment]
@annulen, right, that mobile them was brought before…
It should be fine to have just it I guess.

What about these "variations"? Just these two predefined?
I'm wondering if it could be exploited to allow to apply different styles by allowing to subclass it and provide more variations.

**Comment by annulen** [comment]
>What about these "variations"? Just these two predefined?

Each Page now has (at least) 2 flags affecting style: `useSystemAppearance` and `useDarkAppearance`. Implementation of RenderTheme can take these flags into account and return different values, but it doesn't have a place to store entire per-page QStyle (being global singleton). Though it's of course possible to bolt QStyle into Page under #if PLATFORM(QT)...

**Comment by Emdek** [comment]
@annulen, OK, one thing is for sure, it doesn't seen to make sense to backport it to 5.212.

**Comment by annulen** [comment]
Yet another style flag of Page is `useElevatedUserInterfaceLevel` which comes from https://developer.apple.com/documentation/uikit/uiuserinterfacelevel/elevated, but that may be too Mac-specific

**Comment by annulen** [comment]
About backporting I've meant `@media (prefers-color-scheme: dark)` thing - however it's not really trivial to backport, and we have no means to deduce if color scheme is light or dark from existing QStyle, so it can be set by user manually only. But if we migrate t built-in style, we can make things consistent easily.

**Comment by Emdek** [comment]
@annulen, I'm fine with built-in style, but some way to customize it would be great (doesn't have to be `QStyle`).

**Comment by annulen** [comment]
From the top of my head I can see it having "main color" as an option. Any approach involving subclassing would be inconvenient for WK2, as it would work only from WebProcess plugin

**Comment by Emdek** [comment]
@annulen, uhm, main color is always something.
Main for each mode (light vs dark)?

**Comment by annulen** [comment]
Dark theme may be as simple as color inversion of default

**Comment by Emdek** [comment]
@annulen, yeah, but these "simple inversions" are usually less reliable than random Qt styles using predefined light or dark palettes. ;-)

**Comment by annulen** [comment]
See https://github.com/qtwebkit/qtwebkit/blob/qtwebkit-dev-wip/Source/WebCore/platform/qt/RenderThemeQtMobile.cpp

**Comment by annulen** [comment]
>RenderTheme was changed from being per-page object to singleton...

See also 19905ae697a
